### ISSUE TITLE
Question about the KL Divergence

### ISSUE THREAD

**Issue by deadmarston** [Tom94/practical-path-guiding#11]
Hi Tom,

Thanks for your great work.
Regarding your gradient descent for KL Divergence, I wonder is there any possibility that your optimized bsdfSamplingFraction will fall into a local optimization because the whole sphere domain share a same bsdfSamplingFraction?

**Comment by Tom94** [comment]
Hi deadmarston,

the solution is definitely suboptimal because of the lack of dependence on incident direction as well as the averaging within the spatial SD-tree cell.

Alleviating either of these two limitations would make the learned BSDF sampling fraction more accurate.

Regarding local minima: applying SGD on the spatio-directionally averaged gradients amounts to finding an optimum of the spatio-directionally averaged loss function... so I believe you may be right about the averaging introducing local optima. Averaging multiple loss functions with a single global optimum can absolutely lead to multiple local optima, but I'm not able to make a definitive statement whether this'll happen in the specific use of the KL divergence that we have here (and under which conditions).

Cheers!

**Comment by deadmarston** [comment]
Hi Tom,

Thank you for your valueable reply:)

Cheers!
